### DISCUSSION TITLE
Testnet - Guide to Setup Acala/Karura Testnet Nodes

### DISCUSSION THREAD

**Discussion by andreclaro** [AcalaNetwork/Acala#2769] [0. General]
Hello, 

I would like to setup testnet nodes for Acala and Karura. Can someone help with which flags/configuration should I use?

Thank you!

**Comment by andreclaro** [comment]
Tried to run acala with the following settings (`--chain=acala-dev`):
```
/usr/local/bin/acala \
    --base-path=/var/lib/acala/data \
    --chain=acala-dev \
    --name=XXXX \
    --database=paritydb \
    --pruning=200000 \
    --rpc-cors=all \
    --rpc-external \
    --rpc-port=9933 \
    --rpc-max-connections=500 \
    --port=30333 \
    --execution=wasm \
    --prometheus-external \
    --prometheus-port=961
```

But acala is not getting peers:
```
Jun 13 08:06:39 t-acala-01 acala[73608]: 2024-06-13 08:06:39.748  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #0 (0x9af3…7e64), finalized #0 (0x9af3…7e64), ⬇ 0 ⬆ 0
Jun 13 08:06:44 t-acala-01 acala[73608]: 2024-06-13 08:06:44.748  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #0 (0x9af3…7e64), finalized #0 (0x9af3…7e64), ⬇ 0 ⬆ 0
Jun 13 08:06:49 t-acala-01 acala[73608]: 2024-06-13 08:06:49.748  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #0 (0x9af3…7e64), finalized #0 (0x9af3…7e64), ⬇ 0 ⬆ 0
Jun 13 08:06:54 t-acala-01 acala[73608]: 2024-06-13 08:06:54.749  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #0 (0x9af3…7e64), finalized #0 (0x9af3…7e64), ⬇ 0 ⬆ 0
:Jun 13 08:06:59 t-acala-01 acala[73608]: 2024-06-13 08:06:59.749  INFO tokio-runtime-worker substrate: 💤 Idle (0 peers), best: #0 (0x9af3…7e64), finalized #0 (0x9af3…7e64), ⬇ 0 ⬆ 0
```

**Reply by zjb0807** [reply]
acala-dev is a solo mode, which does not require connecting to peers. Blocks are generated when there are transactions.

**Reply by andreclaro** [reply]
So there is no real public testnet, where we could connect our node to?

This [Acala testnet](https://evmdocs.acala.network/network/network-configuration#acala-testnet) is also a solo network without peers?

**Reply by zjb0807** [reply]
This is a testnet with peerchain and relaychain, but it has not been updated for a long time. 
What is the scenario for you to start the testnet? The recommended way now is to fork the mainnet with [chopsticks](https://github.com/AcalaNetwork/chopsticks)

You can run it directly from [polkadot.js.org](https://polkadot.js.org/apps/#/explorer) 
<img width="554" alt="image" src="https://github.com/AcalaNetwork/Acala/assets/16496415/423f19d2-e284-4122-89f6-9f80288df66c">

**Reply by andreclaro** [reply]
We want to run testnet nodes for Acala and Karura as part of the wormhole testnet environment.

<img width="747" alt="image" src="https://github.com/AcalaNetwork/Acala/assets/33039405/03f89606-c928-467e-848a-8ad2415c8cc6">

https://wormhole-foundation.github.io/wormhole-dashboard/#/?endpoint=Testnet

**Reply by andreclaro** [reply]
We tried to use `wss://eth-rpc-acala-testnet.aca-staging.network`, but we are having issues.


```
websocat wss://eth-rpc-acala-testnet.aca-staging.network

{"method":"eth_getBlockByNumber","params":["latest",false],"id":1,"jsonrpc":"2.0"}
{"id":1,"jsonrpc":"2.0","error":{"code":6969,"message":"Error: WebSocket is not connected"}}

{"method":"eth_getBlockByNumber","params":["safe",false],"id":1,"jsonrpc":"2.0"}
{"id":1,"jsonrpc":"2.0","error":{"code":6969,"message":"Error: WebSocket is not connected"}}

{"method":"eth_getBlockByNumber","params":["finalized",false],"id":1,"jsonrpc":"2.0"}
{"id":1,"jsonrpc":"2.0","error":{"code":6969,"message":"Error: WebSocket is not connected"}}
```

**Reply by zjb0807** [reply]
@xlc Do you have any suggestions? 
Should we continue to maintain `acala-local`/`karura-local` to support wormhole testnet?

**Reply by andreclaro** [reply]
@xlc @zjb0807,  alternatively, can provide us a working web socket endpoints for your acala and karura testnet nodes? Thanks

**Reply by xlc** [reply]
There were some issue on our hosted testnets and we are fixing it. You should be able to use them after the issue is fixed.

**Reply by andreclaro** [reply]
Thanks a million, @xlc 🙏🏽  .... Please let me know when fixed. Cheers

**Reply by andreclaro** [reply]
Btw, FYI, we will need to use your testnet RPC nodes because the wormhole testnet contracts were deployed to your testnet networks:
- https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/env/.env.acala.testnet
- https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/env/.env.karura.testnet

**Reply by xlc** [reply]
The Karura testnet is fixed. Working on fix the Acala one now.

**Reply by zjb0807** [reply]
I'm still working on the Karura testnet. The relaychain needs to be upgraded

**Reply by andreclaro** [reply]
I can confirm the Karura testnet endpoint (websocket) is working for us.

Let me know when Acala testnet endpoint is fixed so we can also validate it. 

Thanks a million! Cheers

**Reply by zjb0807** [reply]
Karura and Acala testnet can't generate blocks. Can I reset the testnet and then redeploy the wormhole contract?

**Reply by andreclaro** [reply]
Let me double check.

**Reply by andreclaro** [reply]
is there any other alternative to recover these testnet networks? if not, please go ahead and let me know when we can re-deploy the contracts.

**Reply by andreclaro** [reply]
hello @zjb0807, is there any update on Acala and Karura testnet networks? Thanks

**Reply by zjb0807** [reply]
The testnet has been restored, but there is still some configuration work to do. I will provide it to you ASAP.

**Reply by zjb0807** [reply]
acala eth rpc: https://eth-rpc-acala-testnet.aca-staging.network
acala blockscout: https://blockscout.acala-testnet.aca-staging.network


karura eth rpc: https://eth-rpc-karura-testnet.aca-staging.network
karura blockscout: https://blockscout.karura-testnet.aca-staging.network

I sent 1000 ACA/KAR to the 0x8F26A0025dcCc6Cfc07A7d38756280a10E295ad7, please update the testnet and re-deploy the contracts.

We have updated the node so that it is no longer necessary to specify the gas-price parameter.
https://github.com/wormhole-foundation/wormhole/blob/627faa7e694eeadc72615873cdbf9985385c888d/ethereum/env/.env.acala.testnet#L4-L5

**Reply by andreclaro** [reply]
Hello @zjb0807, Thanks a million! We will get back to you when contracts are deployed. Cheers!

**Reply by andreclaro** [reply]
@zjb0807 , can you confirm the karura node is operating correctly? I'm unable to deploy the wormhole core contract... I'm getting `server returned an error response: error code -32603: execution reverted:`

It worked for acala: https://blockscout.acala-testnet.aca-staging.network/address/0x64fb09E405D2043ed7785a29E296C766D56F2056

**Reply by zjb0807** [reply]
@andreclaro Do you happen to have more details?  Is there a raw transaction that I can broadcast to the network again?
I see all transactions are successful on the karura testnet. 
https://blockscout.karura-testnet.aca-staging.network/txs?tab=validated

**Reply by andreclaro** [reply]
hi @zjb0807 , we are trying to understand what went wrong and I will share more with you soon. Thanks

**Reply by andreclaro** [reply]
Still getting `server returned an error response: error code -32603: execution reverted:` for the Karura contracts deployment.

```
MNEMONIC=... ./sh/deployCoreBridge.sh

[⠒] Compiling...
No files changed, compilation skipped
Traces:
  [3581426] DeployCore::run([0x13947Bd48b18E53fdAeEe77F3473391aC727C638], 11, 1, 0x0000000000000000000000000000000000000000000000000000000000000004, 596)
    ├─ [0] VM::startBroadcast()
    │   └─ ← [Return]
    ├─ [2907577] → new Implementation@0x33938BC90E51450aa7434Beb27299529b3AF29d1
    │   └─ ← [Return] 14522 bytes of code
    ├─ [274512] → new Setup@0x28136c299C7C51CB331f89A17D9Bd6266B613dA6
    │   └─ ← [Return] 1371 bytes of code
    ├─ [293821] → new Wormhole@0x798163b565462846d7411444DB96F5E03B6D4Dcc
    │   ├─ emit Upgraded(implementation: Setup: [0x28136c299C7C51CB331f89A17D9Bd6266B613dA6])
    │   ├─ [130616] Setup::setup(Implementation: [0x33938BC90E51450aa7434Beb27299529b3AF29d1], [0x13947Bd48b18E53fdAeEe77F3473391aC727C638], 11, 1, 0x0000000000000000000000000000000000000000000000000000000000000004, 596) [delegatecall]
    │   │   ├─ emit Upgraded(implementation: Implementation: [0x33938BC90E51450aa7434Beb27299529b3AF29d1])
    │   │   └─ ← [Stop]
    │   └─ ← [Return] 680 bytes of code
    ├─ [0] VM::stopBroadcast()
    │   └─ ← [Return]
    └─ ← [Return] Wormhole: [0x798163b565462846d7411444DB96F5E03B6D4Dcc], Setup: [0x28136c299C7C51CB331f89A17D9Bd6266B613dA6], Implementation: [0x33938BC90E51450aa7434Beb27299529b3AF29d1]


Script ran successfully.

== Return ==
deployedAddress: address 0x798163b565462846d7411444DB96F5E03B6D4Dcc
setupAddress: address 0x28136c299C7C51CB331f89A17D9Bd6266B613dA6
implAddress: address 0x33938BC90E51450aa7434Beb27299529b3AF29d1

## Setting up 1 EVM.
==========================
Simulated On-chain Traces:

  [2907577] → new Implementation@0x33938BC90E51450aa7434Beb27299529b3AF29d1
    └─ ← [Return] 14522 bytes of code

  [274512] → new Setup@0x28136c299C7C51CB331f89A17D9Bd6266B613dA6
    └─ ← [Return] 1371 bytes of code

  [293821] → new Wormhole@0x798163b565462846d7411444DB96F5E03B6D4Dcc
    ├─ emit Upgraded(implementation: Setup: [0x28136c299C7C51CB331f89A17D9Bd6266B613dA6])
    ├─ [130616] Setup::setup(Implementation: [0x33938BC90E51450aa7434Beb27299529b3AF29d1], [0x13947Bd48b18E53fdAeEe77F3473391aC727C638], 11, 1, 0x0000000000000000000000000000000000000000000000000000000000000004, 596) [delegatecall]
    │   ├─ emit Upgraded(implementation: Implementation: [0x33938BC90E51450aa7434Beb27299529b3AF29d1])
    │   └─ ← [Stop]
    └─ ← [Return] 680 bytes of code


==========================

Chain 596

Estimated gas price: 100.000082577 gwei

Estimated total gas used for script: 5094119

Estimated amount required: 0.509412320657064663 ETH

==========================
##
Sending transactions [0 - 2].

Transactions saved to: /Users/andreclaro/Code/01.xLabs/wormhole/ethereum/broadcast/DeployCore.s.sol/596/run-latest.json

Sensitive values saved to: /Users/andreclaro/Code/01.xLabs/wormhole/ethereum/cache/DeployCore.s.sol/596/run-latest.json

Error:
Failed to send transaction

Context:
- server returned an error response: error code -32603: execution reverted:
-- Wormhole Core Addresses --------------------------------------------------
| Setup address                | 0x28136c299C7C51CB331f89A17D9Bd6266B613dA6 |
| Implementation address       | 0x33938BC90E51450aa7434Beb27299529b3AF29d1 |
| Wormhole address             | 0x798163b565462846d7411444DB96F5E03B6D4Dcc |
-----------------------------------------------------------------------------
```

```
cast code -r https://eth-rpc-karura-testnet.aca-staging.network 0x28136c299C7C51CB331f89A17D9Bd6266B613dA6
0x

cast code -r https://eth-rpc-karura-testnet.aca-staging.network 0x33938BC90E51450aa7434Beb27299529b3AF29d1
0x

cast code -r https://eth-rpc-karura-testnet.aca-staging.network 0x798163b565462846d7411444DB96F5E03B6D4Dcc
0x
```

And we can use the verify.sh script to verify the deployment of each contract, such as:
```
./verify -r https://eth-rpc-karura-testnet.aca-staging.network  -c karura  build-forge/Setup.sol/Setup.json 0x28136c299C7C51CB331f89A17D9Bd6266B613dA6

Deployed bytecode of 0x28136c299C7C51CB331f89A17D9Bd6266B613dA6 on karura doesn't match build-forge/Setup.sol/Setup.json
deployed hash:
d0a00c80ae48701de756e7a5a5a59eabb25c71cd60b9cee5a193e895b25b163f
build-forge/Setup.sol/Setup.json hash:
a3a43045633400865799f21a6f2b76d0e7382b6152dfcb25538a95bd41e26618
```


You can try on your side:
1. git clone https://github.com/wormhole-foundation/wormhole
2. cd wormhole/ethereum
3. make build
5. ln -s env/.env.karura.testnet .env. # review env/.env.karura.testnet  file
6. MNEMONIC=ADD_YOUR_WALLET_SECRET ./sh/deployCoreBridge.sh


btw, I have a PR to remove the gas price setting: https://github.com/wormhole-foundation/wormhole/pull/3995/files

This worked for Acala... but it is always failing with Karura..

**Reply by zjb0807** [reply]
Thank you for your reply.  The karura testnet still has some problems. I'll try your steps when the network works

**Reply by andreclaro** [reply]
I tried again two times and now I got the following errors:
- 1st: "server returned an error response: error code -32603: 1012: Transaction is temporarily banned"
- 2nd: "server returned an error response: error code 6969: Error: 1014: Priority is too low: (147 vs 0): The transaction has too low priority to replace another transaction already in the pool."

now I'm always getting the 2nd message. 🤔

**Reply by zjb0807** [reply]
The network cannot generate blocks, so the transaction cannot be completed now

**Comment by shunjizhan** [comment]
hi @andreclaro, after I attested [reth](https://sepolia.etherscan.io/token/0xa7a16438A3B190BEd16364cF18ed1CbA94693010) from sepolia to acala testnet, on [transfer page](http://localhost:3000/?cluster=testnet#/transfer) it still shows `Target asset unavailable`. 
Could you please double check if the contracts are configured correctly? Thanks

The UI I am using is [this PR](https://github.com/wormhole-foundation/example-token-bridge-ui/pull/69), with fixed acala rpc url 

links:
- [attest tx on sepolia](https://sepolia.etherscan.io/tx/0x0f7de35f6fb2771bf1a4967de4101ab111a93c8b9c8b609ff52b12456cf09c6b)
- [register tx on acala testnet](https://blockscout.acala-testnet.aca-staging.network/tx/0x16a9f43e10865d1d6f1d2e13298d1072134f13a3cf87fb92c57c824dc114c53b)

![Screenshot 2024-07-01 at 12 19 16](https://github.com/AcalaNetwork/Acala/assets/16548786/80a3ec4c-f4a8-4bf7-a64d-305fb2818088)
![Screenshot 2024-07-01 at 12 26 04](https://github.com/AcalaNetwork/Acala/assets/16548786/5f6586ce-164f-414b-9c2e-0a7885aacfb0)

**Comment by andreclaro** [comment]
We are still working on it... I will let you know when it is completed.

**Reply by shunjizhan** [reply]
thanks a lot!

**Comment by andreclaro** [comment]
https://blockscout.karura-testnet.aca-staging.network is down...

**Reply by zjb0807** [reply]
It works now. 
There were some problems with the test network, but now it works

**Reply by andreclaro** [reply]
again down 😢

**Reply by zjb0807** [reply]
I'm working on it :sweat_smile:

**Reply by andreclaro** [reply]
any news? thanks

**Reply by zjb0807** [reply]
I have the same problem on the acala testnet. 
I am still working on it and I am not sure if it will reset the network again :slightly_frowning_face:

**Reply by andreclaro** [reply]
Feel fee to make sure what is required to fix the issues on both networks. I will redeploy te contracts on Acala if required.

**Reply by andreclaro** [reply]
@zjb0807 , is there any update about acala and karura testnet?

**Reply by andreclaro** [reply]
Any update @zjb0807 ?

**Reply by zjb0807** [reply]
The karura testnet has been reset. I sent 1000 KAR to `0x8F26A0025dcCc6Cfc07A7d38756280a10E295ad7`.

I'm debugging the reverted TX. Do you need the test coin? Please tell me the address.

**Reply by zjb0807** [reply]
@andreclaro please use `FORGE_ARGS="--slow --legacy"` to deploy the contracts. It'll work

**Reply by andreclaro** [reply]
Thanks @zjb0807 , Can you send KAR to my testnet wallet please? Address: 0x57B3d968f3121585855C175Cae2032D9141f23e5

Is the Acala testnet now stable?

**Reply by zjb0807** [reply]
https://blockscout.karura-testnet.aca-staging.network/tx/0x89b24f7c15b7356bc85d7e20c0bd4bfba20680651995306d2728252ccc57e043

Both Acala and Karura should be stable now.

**Reply by andreclaro** [reply]
I was able to deploy Karura contracts. Thanks!

**Reply by andreclaro** [reply]
Hi @zjb0807 , while trying to upgrade the smart contract using our script, I go the following errors on both acala and karura testnet:

```
Error:
Failed to send transaction

Context:
- server returned an error response: error code 6969: Error: internal JSON-RPC error [evm.OutOfStorage: ] [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (from="0x57B3d968f3121585855C175Cae2032D9141f23e5", to="0xe157115ef34c93145Fec2FE53706846853B07F42", gasLimit="11280000", storageLimit="2", value="0", data="0xa5799f93000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000c001000000000100c0c64027377f27bb3ef676f2fd9c63a30a6d1bf904851656cad635601425191017308e3d0e605364c3a1a5305b722fd150bde1f869b74a8cbb686774c41a7c880100000001000000010001000000000000000000000000000000000000000000000000000000000000000400000000051a364b00000000000000000000000000000000000000000000546f6b656e42726964676501000000013b26409f8aaded3f5ddca184695aa6a0fa829b0c85caf84856324896d214ca98", accessList=undefined, code=CALL_EXCEPTION, version=bodhi.js/providers/2.8.7)
```

Is this issue related with your RPC nodes?


Script: https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/sh/registerAllChainsOnTokenBridge.sh 
- I added ` --legacy` to the forge command since I was initially getting `Failed to get EIP-1559 fees` error.
- The following branch is updated with the latest sdks: https://github.com/wormhole-foundation/wormhole/pull/4064

**Reply by zjb0807** [reply]
@andreclaro how can I reproduce the error?

```
MNEMONIC=... ./sh/registerAllChainsOnTokenBridge.sh testnet karura
```
get the error:
```
0000000000000000000000c7a204bdbfe983fcd8d8e61d02b475d4073ff97e])
    ├─ [0] VM::startBroadcast()
    │   └─ ← [Return]
    └─ ← [Revert] EvmError: Revert


Error:
script failed: <empty revert data>
```

**Reply by andreclaro** [reply]
Not sure if this must be done with the same wallet (MNEMONIC) used to deploy the contract...

But you need to use this PR, which has the last sdk releases: https://github.com/wormhole-foundation/wormhole/pull/4064

I added `--legacy to the forge command` https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/sh/registerAllChainsOnTokenBridge.sh#L101-L105

Then:
```
cd ethereum
make build
MNEMONIC=... ./sh/registerAllChainsOnTokenBridge.sh testnet karura
```

**Reply by zjb0807** [reply]
<img width="682" alt="image" src="https://github.com/user-attachments/assets/68f216ad-9359-46cf-92c6-fb07b514f8a3">

I got the error after adding `--skip-simulation` to the forge command

**Reply by zjb0807** [reply]
should work with https://github.com/foundry-rs/foundry/pull/8625

**Reply by zjb0807** [reply]
The PR was merged. Please use the latest forge to retry.

**Comment by andreclaro** [comment]
Hello @zjb0807  , hope you are doing well!

Karura testnet has stalled at block `369593` .

Explorer is also not working: https://blockscout.karura-testnet.aca-staging.network

Can you take a look please?

Thanks!

**Reply by zjb0807** [reply]
The disk was full, but it's working now. Please check it.

**Reply by andreclaro** [reply]
Thanks a million @zjb0807

**Reply by andreclaro** [reply]
Hi @zjb0807 , hope you are doing well! 

is there any issue with Acala testnet endpoint? Thanks

**Reply by zjb0807** [reply]
The disk was full, so I cleaned it. It's working.

**Comment by zjb0807** [comment]
@andreclaro 
Hello, we will be stopping maintenance of the Acala/Karura testnet. 
Please help update Guardians. Thanks.

**Reply by andreclaro** [reply]
Thanks a million @zjb0807 for all your help and support!
